libjulia/interpreter/Interpreter.cpp

```diff
@@ -0,0 +1,225 @@
+		{
+			FunctionDefinition const& funDef = boost::get<FunctionDefinition>(statement);
+			m_functions[funDef.name] = &funDef;
+		}
+	ASTWalker::operator()(_block);
+	closeScope();
+}
+
+u256 Interpreter::evaluate(Expression const& _expression)
+{
+	ExpressionEvaluator ev(m_state, m_variables, m_functions);
+	ev.visit(_expression);
+	return ev.value();
+}
+
+vector<u256> Interpreter::evaluateMulti(Expression const& _expression)
+{
+	ExpressionEvaluator ev(m_state, m_variables, m_functions);
+	ev.visit(_expression);
+	return ev.values();
+}
+
+void Interpreter::closeScope()
+{
+	for (auto const& var: m_scopes.back())
+	{
+		solAssert(m_variables.erase(var) + m_functions.erase(var) == 1, "");
+	}
+}
+
+void ExpressionEvaluator::operator()(Literal const& _literal)
+{
+	using solidity::assembly::LiteralKind;
+	switch (_literal.kind)
+	{
+	case LiteralKind::Boolean:
+		solAssert(_literal.value == "true" || _literal.value == "false", "");
+		setValue(_literal.value == "true" ? 1 : 0);
+		break;
+	case LiteralKind::Number:
+		setValue(u256(_literal.value));
+		break;
+	case LiteralKind::String:
+		solAssert(_literal.value.size() <= 32, "");
+		setValue(u256(h256(_literal.value, h256::FromBinary, h256::AlignLeft)));
+		break;
+	}
+}
+
+void ExpressionEvaluator::operator()(Identifier const& _identifier)
+{
+	solAssert(m_variables.count(_identifier.name), "");
+	setValue(m_variables.at(_identifier.name));
+}
+
+void ExpressionEvaluator::operator()(FunctionalInstruction const& _instr)
+{
+	evaluateArgs(_instr.arguments);
+	EVMInstructionInterpreter interpreter(m_state);
+	// The instruction might also return nothing, but it does not
+	// hurt to set the value in that case.
+	setValue(interpreter.eval(_instr.instruction, values()));
+}
+
+void ExpressionEvaluator::operator()(FunctionCall const& _funCall)
+{
+	solAssert(m_functions.count(_funCall.functionName.name), "");
+	evaluateArgs(_funCall.arguments);
+
+	FunctionDefinition const& fun = *m_functions.at(_funCall.functionName.name);
+	solAssert(m_values.size() == fun.parameters.size(), "");
+	map<string, u256> arguments;
+	for (size_t i = 0; i < fun.parameters.size(); ++i)
+		arguments[fun.parameters.at(i).name] = m_values.at(i);
+	for (size_t i = 0; i < fun.returnVariables.size(); ++i)
+		arguments[fun.returnVariables.at(i).name] = 0;
+
+	// TODO function name lookup could be a little more efficient,
+	// we have to copy the list here.
+	Interpreter interpreter(m_state, arguments, m_functions);
+	interpreter(fun.body);
+
+	m_values.clear();
+	for (auto const& retVar: fun.returnVariables)
+		m_values.emplace_back(interpreter.valueOfVariable(retVar.name));
+}
+
+u256 ExpressionEvaluator::value() const
+{
+	solAssert(m_values.size() == 1, "");
+	return m_values.front();
+}
+
+void ExpressionEvaluator::setValue(u256 _value)
+{
+	m_values.clear();
+	m_values.emplace_back(std::move(_value));
+}
+
+void ExpressionEvaluator::evaluateArgs(vector<Expression> const& _expr)
+{
+	vector<u256> values;
```

Comment:
Why do you need this intermediate variable (even if it is moved). Cannot you just reset `m_values` and push directly there?